File: programmers/ForKaKAO/2020/level1/키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    
    dic = {
        1: [0, 0], 2: [0, 1], 3: [0, 2],
        4: [1, 0], 5: [1, 1], 6: [1, 2],
        7: [2, 0], 8: [2, 1], 9: [2, 2],
        '*': [3, 0], 0: [3, 1], '#': [3, 2]
    }
    
    lpos = dic['*']
    rpos = dic['#']
    
    for num in numbers:
        now = dic[num]
        
        if num in [1, 4, 7]:
            answer += 'L'
            lpos = dic[num]
        elif num in [3, 6, 9]:
            answer += 'R'
            rpos = dic[num]
        else:
            ltmp = rtmp = 0
            for a, b, c in zip(lpos, rpos, now):
                ltmp += abs(a - c)
                rtmp += abs(b - c)
            
            if ltmp < rtmp:
                answer += 'L'
                lpos = dic[num]
            elif ltmp > rtmp:
                answer += 'R'
                rpos = dic[num]
            else:
                if hand == "left":
                    answer += 'L'
                    lpos = dic[num]
                else:
                    answer += 'R'
                    rpos = dic[num]
                
    return answer

File: programmers/ForKaKAO/2020/level1/키패드 누르기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int getDist(pair<int, int> h, pair<int, int> n) {
    return abs(n.first - h.first) + abs(n.second - h.second);
}

string solution(vector<int> numbers, string hand) {
    string answer = "";
    vector<pair<int, int>> v[10];
    
    v[0].push_back(make_pair(3, 1));
    int dial = 1;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            v[dial++].push_back(make_pair(i, j));
        }
    }

    pair<int, int> lpos = { 3, 0 };
    pair<int, int> rpos = { 3, 2 };
    for (int i = 0; i < numbers.size(); i++) {
        int num = numbers[i];
        pair<int, int> npos = make_pair(v[num].front().first, v[num].front().second);
        
        if (num == 1 || num == 4 || num == 7) {
            answer += "L";
            lpos = npos;
        } else if (num == 3 || num == 6 || num == 9) {
            answer += "R";
            rpos = npos;
        } else {
            int ldiff = getDist(lpos, npos);
            int rdiff = getDist(rpos, npos);
            if (ldiff < rdiff) {
                answer += "L";
                lpos = npos;
            } else if (ldiff > rdiff) {
                answer += "R";
                rpos = npos;
            } else {
                if (hand.compare("right") == 0) {
                    answer += "R";
                    rpos = npos;
                } else {
                    answer += "L";
                    lpos = npos;
                }
            }
        }
    }
    
    return answer;
}